pkg/queue/concurrency_state_test.go

```diff
@@ -17,116 +17,127 @@ limitations under the License.
-				paused.Inc()
-			}
+	handler := func(w http.ResponseWriter, r *http.Request) {}
+	logger := ltesting.TestLogger(t)
+	h := ConcurrencyStateHandler(logger, http.HandlerFunc(handler), func() { paused.Inc() }, func() { resumed.Inc() })
 
-			resumed := atomic.NewInt64(0)
-			resume := func() {
-				resumed.Inc()
-			}
+	h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", "http://target", nil))
+	if got, want := pollFor(paused, 1), int64(1); got != want {
+		t.Errorf("Pause was called %d times, want %d times", got, want)
+	}
 
-			delegated := atomic.NewInt64(0)
-			delegate := func(w http.ResponseWriter, r *http.Request) {
-				wait, err := strconv.Atoi(r.Header.Get("wait"))
-				if err != nil {
-					panic(err)
-				}
+	if got, want := pollFor(resumed, 1), int64(1); got != want {
+		t.Errorf("Resume was called %d times, want %d times", got, want)
+	}
 
-				time.Sleep(time.Duration(wait))
-				delegated.Inc()
-			}
+	h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", "http://target", nil))
+	if got, want := pollFor(paused, 2), int64(2); got != want {
+		t.Errorf("Pause was called %d times, want %d times", got, want)
+	}
 
-			h := ConcurrencyStateHandler(logger, http.HandlerFunc(delegate), pause, resume)
-
-			var wg sync.WaitGroup
-			wg.Add(len(tt.events))
-			for delay, length := range tt.events {
-				length := length
-				time.AfterFunc(delay, func() {
-					w := httptest.NewRecorder()
-					r := httptest.NewRequest("GET", "http://target", nil)
-					r.Header.Set("wait", strconv.FormatInt(int64(length), 10))
-					h.ServeHTTP(w, r)
-					wg.Done()
-				})
-			}
+	if got, want := pollFor(resumed, 2), int64(2); got != want {
+		t.Errorf("Resume was called %d times, want %d times", got, want)
+	}
+}
 
-			wg.Wait()
-			// Allow last update to finish (otherwise values are off, though this doesn't show
-			// as a race condition when running `go test -race `
-			// TODO Less hacky fix for this
-			time.Sleep(100 * time.Microsecond)
+func TestConcurrencyStateHandlerParallelSubsumed(t *testing.T) {
+	paused := atomic.NewInt64(0)
+	resumed := atomic.NewInt64(0)
 
-			if got, want := paused.Load(), tt.pauses; got != want {
-				t.Errorf("expected to be paused %d times, but was paused %d times", want, got)
-			}
+	req1 := make(chan struct{})
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("req") == "1" {
+			req1 <- struct{}{} // to know it's here.
+			req1 <- struct{}{} // to make it wait.
+		}
+	}
+	logger := ltesting.TestLogger(t)
+	h := ConcurrencyStateHandler(logger, http.HandlerFunc(handler), func() { paused.Inc() }, func() { resumed.Inc() })
 
-			if got, want := delegated.Load(), int64(len(tt.events)); got != want {
-				t.Errorf("expected to be delegated %d times, but delegated %d times", want, got)
-			}
+	go func() {
+		defer func() { req1 <- struct{}{} }()
+		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", fmt.Sprintf("http://target?req=%d", 1), nil))
```

Comment:
do we need the sprintfs here? (I'm totally fine if you think they make things clearer, though I think I'd personally lean towards dropping them)

Reply:
dropping favor of what?

Reply:
ah, nevermind, sure we can drop them. Artifact from another impl.